service/storage_service.cc

```diff
@@ -1419,29 +1431,43 @@ class topology_coordinator {
             }
         });
 
-        if (needs_barrier) {
-            guard = co_await global_tablet_token_metadata_barrier(std::move(guard));
-        }
-
         // In order to keep the cluster saturated, ask the load balancer for more transitions.
         // Unless there is a pending topology change operation.
+        auto ts = guard.write_timestamp();
         auto [preempt, new_guard] = should_preempt_balancing(std::move(guard));
         guard = std::move(new_guard);
+        if (ts != guard.write_timestamp()) {
+            // We rely on the fact that should_preempt_balancing() does not release the guard
+            // so that tablet metadata reading and updates are atomic.
+            on_internal_error(slogger, "should_preempt_balancing() retook the guard");
+        }
```

Comment:
Yes, I think that would be better. That would involve more refactoring because get_node_to_work_on_opt() want to own the guard. I hoped to avoid this refactoring now.

Reply:
We can schedule the refactoring for later (but add a fixme)